storage/range.go

```diff
@@ -1643,9 +1651,32 @@ func (r *Range) LastIndex() (uint64, error) {
 	return atomic.LoadUint64(&r.lastIndex), nil
 }
 
+func (r *Range) raftTruncatedState() (proto.RaftTruncatedState, error) {
+	ts := proto.RaftTruncatedState{}
+	ok, err := engine.MVCCGetProto(r.rm.Engine(), engine.RaftTruncatedStateKey(r.Desc().RaftID),
+		proto.ZeroTimestamp, nil, &ts)
+	if err != nil {
+		return ts, err
+	}
+	if !ok {
+		if len(r.Desc().EndKey) > 0 {
```

Comment:
Might make sense to move the logic for "len(r.Desc().EndKey) > 0" into a new method on Range:

    func (r *Range) isUninitialized() bool {
      return len(r.Desc().EndKey) > 0
    }